wallet-new/docs/updates.md

```diff
@@ -0,0 +1,656 @@
+        , wsuVotesFor        :: !Int
+        , wsuVotesAgainst    :: !Int
+        , wsuPositiveStake   :: !CCoin
+        , wsuNegativeStake   :: !CCoin
+        , wsuSeverity        :: !UpdateSeverity
+        -- ^ The severity of the update (i.e. optional or mandatory).
+        , wsuUpdateState     :: !UpdateState
+        -- ^ The state of this update from users point of view.
+        , wsuSha256Checksum  :: !Sha256Checksum
+        -- ^ The checksum expected for this installer. It will probably
+        -- need to be passed as part of the `UpdateProposal` (maybe in an attribute?),
+        -- or we could reuse the hash which is part of the name, if possible.
+        } deriving (Eq, Show, Generic, Typeable)
+
+where `UpdateSeverity` is defined as an enumeration indicating how compulsory this update is:
+
+    data UpdateSeverity = Mandatory
+                        | Optional
+
+For hard forks updates (or security-critical issues) we might want to set the severity
+to `Mandatory`, for soft forks and other non-critical updates we can label an update as `Optional`.
+
+-   `Optional` updates can be rejected by the user, as well as applied to will at a later stage;
+-   `Mandatory` updates will need to be applied immediately and won't allow the user to proceed
+    in using Daedalus.
+
+On the other hand, `wsuUpdateState` records the users' decision towards the update:
+
+    data UpdateState = Applied
+                     -- ^ The update has been fully applied.
+                     | Downloaded !Sha256Checksum !FilePath
+                     -- ^ The update has been downloaded but not applied yet. A Checksum
+                     -- can be stored to verify the integrity of the download.
+                     | Skipped
+                     -- ^ The update has been skipped by the user.
+                     | Proposed
+                     -- ^ The update has been just proposed and is waiting for users' feedback.
+
+Note how we could add an intermediate `Downloading` state, but this would complicate state-handling
+in case a user hard-quits Daedalus whilst the upgrade is being downloaded. In this case, we would
+need to catch the exception and update the state back to `Proposed`, but that case be more error
+prone than not updating the state in the first place. Having the state back to `Proposed` is
+important so that the wallet is left in a clean state and upon re-launched Daedalus the update can
+be downloaded again with no harm being done.
+
+## Determining the severity<a id="sec-5-2" name="sec-5-2"></a>
+
+For the proposed schema to succeed, we would need a function like the following:
+
+    updateSeverity :: UpdateProposal -> UpdateSeverity
+
+Such function would gather evidence from an `UpdateProposal` that this is a non-negotiable update.
+
+Note how we don't need to commit just yet whether or not we will treat hard-forks as `Mandatory`, as
+there are good arguments for people wanting to stick with an old version of the chain.
```

Comment:
So you say that we _will_ stop Daedalus from working if the update is mandatory, but mandatory updates would be only super-severe security ones? Okay, that makes sense, though I think that notion of mandatory update should be discussed more broadly from both philosophical and technical perspectives (e. g. I believe that voting should work differently for super-urgent mandatory updates, and AFAIK we don't have any ideas about it on Core side yet).

Reply:
I agree. In principle, this proposal could start with everything marked as `Optional` and we could offload the final decision with the team at a later stage.